add `lexproduct`, lexicographic-order product iterator

JeffBezanson opened the pull request.

JeffBezanson committed:
add `lexproduct`, lexicographic-order product iterator

tkelman commented on base/iterator.jl:
````diff
@@ -555,7 +562,7 @@ a tuple whose `i`th element comes from the `i`th argument iterator. The first it
 
 ```jldoctest
-julia> collect(product(1:2,3:5))
+julia> vec(collect(product(1:2,3:5)))
````
run `make docs` and commit the result

JeffBezanson replied:
Ok, I'll do that. I'll go with `IterTools` and see what people think.

JeffBezanson committed:
return `false` for deprecated bindings in `is_exported_from_stdlib`

JeffBezanson committed:
add `IterTools` module around extra iterator functionality

JeffBezanson committed:
if a binding exists in multiple `using`s, prefer a non-deprecated one

This removes a spurious warning if Base.X is deprecated to Mod.X,
and one does `using Mod`. `X` is available from both `Base` and
`Mod`, but we asked for the non-deprecated one so there should
be no warning.

tkelman commented:
I assume this is something you're planning on using for some other feature, which is why it needs to be here rather than Iterators.jl?

JeffBezanson commented:
Not necessarily, they just seem to go together. It also reuses some internal types and functions in Base defined for `product`, so it's much easier to implement here.
